PSICover/PSICover.cs

```diff
@@ -163,31 +165,112 @@ class Analyzer {
          var code = File.ReadAllLines (file);
          for (int i = 0; i < code.Length; i++)
             code[i] = code[i].Replace ('<', '\u00ab').Replace ('>', '\u00bb');
+         int hitCount = 0;
          foreach (var block in blocks) {
             bool hit = hits[block.Id] > 0;
-            string tag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
-            code[block.ELine] = code[block.ELine].Insert (block.ECol, "</span>");
-            code[block.SLine] = code[block.SLine].Insert (block.SCol, tag);
+            if (hit) hitCount++;
+            // html elements - single line block
+            string tooltipTitle = $"{(hit ? hits[block.Id] + (hits[block.Id] > 1 ? " hits" : " hit") : "Code block not hit")}";
+            string titleAttrib = $"data-title=\"{tooltipTitle}\"";
+            string samelineTag = $"<span class=\"{(hit ? "hit tooltip animate" : "unhit tooltip animate")}\" {titleAttrib}>";
+            string closeTag = "</span>";
+            // html elements - multi line block
+            string insideMultilineTag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
+            string multilineTag = $"<span class=\"tooltip animate\" {titleAttrib}>{insideMultilineTag}";
+            // adding html tags to all the blocks
+            for (int lastLine = block.ELine; lastLine >= block.SLine; lastLine--) {
+               if (block.SLine == block.ELine) { // for same line blocks
+                  code[block.ELine] = code[block.ELine].Insert (block.ECol, closeTag);
+                  code[block.SLine] = code[block.SLine].Insert (block.SCol, samelineTag);
+               } else if (lastLine == block.ELine) { // for last line's column of multiline block
+                  code[lastLine] = code[lastLine].Insert (block.ECol, closeTag + closeTag);
+                  code[lastLine] = code[lastLine].Insert (FirstValidIdx (code[lastLine]), insideMultilineTag);
+               } else if (lastLine == block.SLine) { // for first line's column of multiline block
+                  code[lastLine] = code[lastLine].Insert (LastValidIdx (code[lastLine]), closeTag);
+                  code[lastLine] = code[lastLine].Insert (block.SCol, multilineTag);
+               } else { // for multi line blocks, w/wo other blocks on same line
+                  if (!string.IsNullOrEmpty (code[lastLine])) {
+                     int sCol = FirstValidIdx (code[lastLine]);
+                     if (sCol != code[lastLine].Length) {
+                        int lCol = LastValidIdx (code[lastLine]);
+                        code[lastLine] = code[lastLine].Insert (lCol, closeTag);
+                        code[lastLine] = code[lastLine].Insert (sCol, insideMultilineTag);
+                     }
+                  }
+               }
+            }

```

Comment:
I think you are using some html tag to highlight the code. Please see my code:
```
            for (int i = block.ELine; i >= block.SLine; i--) {
               int startCol = block.SCol, endCol = block.ECol;
               if (i != block.ELine) endCol = code[i].Length - code[i].Reverse ().TakeWhile (char.IsWhiteSpace).Count ();
               if (i != block.SLine) startCol = code[i].TakeWhile (char.IsWhiteSpace).Count ();
               code[i] = code[i].Insert (endCol, string.Format (endTag, hits[block.Id]));
               code[i] = code[i].Insert (startCol, startTag);
            }
```
Can you refactor your code based on above code? Also, instead of lastline variable you could simply use i

Reply:
Variable name shortened. But I don't know how to refactor the code further. Could you help me out?
FYI, I am using two span for multiline blocks.

Reply:
```
         int cHits = 0;
         foreach (var block in blocks) {
            bool hit = hits[block.Id] > 0;
            if (hit) cHits++;
            // html elements - single line block
            string tooltipTitle = $"{(hit ? hits[block.Id] + (hits[block.Id] > 1 ? " hits" : " hit") : "Code block not hit")}";
            string titleAttrib = $"data-title=\"{tooltipTitle}\"";
            string samelineTag = $"<span class=\"{(hit ? "hit tooltip animate" : "unhit tooltip animate")}\" {titleAttrib}>";
            string closeTag = "</span>";
            // html elements - multi line block
            string insideMultilineTag = $"<span class=\"{(hit ? "hit" : "unhit")}\">";
            string multilineTag = $"<span class=\"tooltip animate\" {titleAttrib}>{insideMultilineTag}";
            // adding html tags to all the blocks
            for (int i = block.ELine; i >= block.SLine; i--) {
               string str = code[i];
               if (string.IsNullOrWhiteSpace (str)) continue;
               bool endLineInMultiLine = block.SLine != block.ELine && i == block.ELine;
               string endTag = endLineInMultiLine ? closeTag + closeTag : closeTag;
               int endCol = i == block.ELine ? block.ECol : LastValidIdx (str);
               code[i] = str.Insert (endCol, endTag);
               int sCol = block.SCol; string sTag = block.SLine == block.ELine ? samelineTag : multilineTag;
               if (i != block.SLine) {
                  sCol = FirstValidIdx (str);
                  sTag = insideMultilineTag;
               }
               code[i] = code[i].Insert (sCol, sTag);
            }
         }
```
Please see above for my implementation.